packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js

```diff
@@ -92,18 +92,45 @@ class UniswapPriceFeed extends PriceFeedInterface {
     // Approximate the first block from which we'll need price data from based on the
     // lookback and twap length:
     const lookbackWindow = this.twapLength + this.historicalLookback;
+    const currentTime = await this.getTime();
+    const earliestLookbackTime = currentTime - lookbackWindow;
     const latestBlockNumber = (await this.web3.eth.getBlock("latest")).number;
-    // Add cushion in case `averageBlockTimeSeconds` overestimates the seconds per block:
-    const lookbackBlocks = Math.ceil((this.bufferBlockPercent * lookbackWindow) / (await averageBlockTimeSeconds()));
-    const earliestBlockNumber = latestBlockNumber - lookbackBlocks;
-    let fromBlock = earliestBlockNumber;
-    let events = await this.uniswap.getPastEvents("Sync", { fromBlock: Math.max(fromBlock, 0) });
-
-    // For low-volume pools, it is possible that there are no Sync events within the lookback window.
-    // To cover these cases, we'll keep looking back until we find a window with a sync event.
-    while (fromBlock >= 0 && events.length === 0) {
-      fromBlock -= lookbackBlocks;
-      events = await this.uniswap.getPastEvents("Sync", { fromBlock: Math.max(fromBlock, 0) });
+    // Add cushion in case `averageBlockTimeSeconds` underestimates the seconds per block:
+    let lookbackBlocks = Math.ceil((this.bufferBlockPercent * lookbackWindow) / (await averageBlockTimeSeconds()));
+
+    let events = []; // Caches sorted events (to keep subsequent event queries as small as possible).
+    let blocks = {}; // Caches blocks (so we don't have to re-query timestamps).
+    let startBlock = Infinity; // Arbitrary initial value > 0.
+
+    // For loop continues until the start block hits 0 or the first event is before the earlest lookback time.
+    for (let i = 0; !(startBlock === 0 || events[0]?.timestamp <= earliestLookbackTime); i++) {
+      // Uses latest unless the events array already has data. If so, it only queries _before_ existing events.
+      const endBlock = events[0] ? events[0].blockNumber - 1 : "latest";
+
+      // By taking larger powers of 2, this doubles the lookback each time.
+      startBlock = Math.max(0, latestBlockNumber - lookbackBlocks * 2 ** i);
+
+      const newEvents = await this._getSortedSyncEvents(startBlock, endBlock).then(newEvents => {
```

Comment:
seeing this does not run in parallel any more, what is the upper bound on how long this `update` function can now take to run, assuming this needs to search until the 0th block and never finds any `sync` events( for example a mis parameterized feed)?

Reply:
Check my math here:
```
avg_block_time = 13.5;
standard_lookback = 7200; // Standard lookback we use with no TWAP.
total_blocks = 14140850; // ~block number in a year
lookback_in_blocks = standard_lookback / avg_block_time = 533.333333333;
iterations = log2(total_blocks / lookback_in_blocks) = 14.6944718176; // Number of times needed to double to reach the total blocks.
 ```

This means it's basically 15 ish loops, which would equate to O(15) serial calls, since no timestamp calls would be required in the average loop since a timestamp call would mean we found a price, terminating the looping.

Does that match your expectation?

Reply:
If we prefer, we can change this from 2 to 3 or even 10, etc if we want to make the ramp up faster. For instance, if we change it to 10, that brings the max iterations down from 15 to 5 ish, but it would mean that the queries would be much more painful for the node to fulfill much more quickly.